kvserver/rangefeed: add more metrics for catch-up

wenyihu6 opened the pull request:
**kvserver/rangefeed: add kv.rangefeed.catchup_scan_blocked_nanos**

This patch adds a new metrics `kv.rangefeed.catchup_scan_blocked_nanos` which
tracks total time spent in rangefeed waiting for the server-side catch-up
semaphore.

Epic: none
Release note (ops change): a new metrics
`kv.rangefeed.catchup_scan_blocked_nanos` can be used to track total time spent
in rangefeed waiting for the server-side catch-up semaphore.


---


**kvclient/rangefeed: add distsender.rangefeed.catchup_scan_blocked_nanos**

This patch adds a new metrics `distsender.rangefeed.catchup_scan_blocked_nanos`
which tracks total time spent in rangefeed waiting for the client-side catch-up
rate limiter.

Epic: none
Release note (ops change): a new metrics
`distsender.rangefeed.catchup_scan_blocked_nanos` can be used to track total time
spent in rangefeed waiting for the client-side catch-up rate limiter.

wenyihu6 committed:
kvserver/rangefeed: add kv.rangefeed.catchup_scan_blocked_nanos

This patch adds a new metrics `kv.rangefeed.catchup_scan_blocked_nanos` which
tracks total time spent in rangefeed waiting for the server-side catch-up
semaphore.

Epic: none
Release note (ops change): a new metrics
`kv.rangefeed.catchup_scan_blocked_nanos` can be used to track total time spent
in rangefeed waiting for the server-side catch-up semaphore.

wenyihu6 committed:
kvclient/rangefeed: add distsender.rangefeed.catchup_scan_blocked_nanos

This patch adds a new metrics `distsender.rangefeed.catchup_scan_blocked_nanos`
which tracks total time spent in rangefeed waiting for the client-side catch-up
rate limiter.

Epic: none
Release note (ops change): a new metrics
`distsender.rangefeed.catchup_scan_blocked_nanos` can be used to track total time
spent in rangefeed waiting for the client-side catch-up rate limiter.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/135625)
<!-- Reviewable:end -->

wenyihu6 commented:
I was considering adding another metric to count the number of rangefeeds being blocked on this semaphore simultaneously - incrementing before starting to acquire the quota and decrementing afterwards. Do you think this would be useful?

stevendanna commented on pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go:
```diff
@@ -251,10 +251,12 @@ func (m *rangefeedMuxer) startSingleRangeFeed(
 func (s *activeMuxRangeFeed) start(ctx context.Context, m *rangefeedMuxer) error {
 	streamID := atomic.AddInt64(&m.seqID, 1)
 
+	startTime := timeutil.Now()
```
Let's use the new crtime.NowMono() for this. It's a little bit more efficient.

```golang
startTime := crtime.NowMono()
// ...
m.metrics.RangefeedCatchUpBlockedNanos.Inc(startTime.Elapsed().Nanoseconds())

wenyihu6 replied:
TIL, done.
<!-- Reviewable comment -ODS669D851L2lMDgMX3:bj9zbl2 -->
<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
Seems reasonable, but I wonder if we want a histogram here. I suppose we could always change it to a histogram later.

wenyihu6 committed:
kvsever/rangefeed: change kv.rangefeed.catchup_scan_blocked_nanos to Histogram

This patch updates kv.rangefeed.catchup_scan_blocked_nanos to use histogram
metrics, enabling more insights with percentiles o better identify outliers.

Release note: none
Epic: none

wenyihu6 committed:
kvclient/rangefeed: change distsender.rangefeed.catchup_scan_blocked_nanos to Histogram

This patch updates distsender.rangefeed.catchup_scan_blocked_nanos to use
histogram metrics, enabling more insights with percentiles o better identify
outliers.

Release note: none
Epic: none

wenyihu6 reviewed (commented):
Good idea, I went ahead and changed it to a histogram in my third and fourth commit. I'm not super confident with the histogram options I chose but went with my intuitions. Let me know your thoughts. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135625#-:-ODS5n5uBZ6flNItzxgv:b-snle8v)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @asg0451, @rharding6373, and @stevendanna)



<!-- Sent from Reviewable.io -->
